### DISCUSSION TITLE
Mismatch between paper and code in setting transmittance below threshold to zero

### DISCUSSION THREAD

**Discussion by dingjr7** [NVlabs/instant-ngp#385] [Q&A]
Dear developers,

I found in appendix C.1 in your paper that "Lastly, we stop ray marching and set the remaining contribution
to zero as soon as the transmittance of the ray drops below 10−4".

However, in src/testbed_nerf.cu:1625
```
	if (src_payload.alive) {
		uint32_t idx = atomicAdd(counter, 1);
		dst_payloads[idx] = src_payload;
		dst_rgba[idx] = src_rgba[i];
	} else if (src_rgba[i].w() > 0.001f) {
		uint32_t idx = atomicAdd(finalCounter, 1);
		dst_final_payloads[idx] = src_payload;
		dst_final_rgba[idx] = src_rgba[i];
	}
```
which is 10^-3 not 10^-4. Hoping to hear from you!

Best regards!

**Comment by Tom94** [comment]
Hi there, that code is not actually the condition for terminating the ray.

The code that you linked checks whether anything was hit at all before the ray died by thresholding the _opacity_ (not the transmittance) with 10^-3. Essentially, it skips compositing entirely for rays that didn't interact appreciably with anything.

The termination criterion is
```cpp
		if (local_rgba.w() > (1.0f - min_alpha)) {
			local_rgba /= local_rgba.w();
			break;
		}
```
from `composite_kernel_nerf`, where `min_alpha` is set to `1e-4` in all the paper results, although the GUI uses `1e-2` by default for slightly higher performance.

That said, the variable `min_alpha` is actually misnamed... it should be called `min_transmittance`. Will fix. :)

Cheers!
